lib/auth/grpcserver.go

```diff
@@ -5204,6 +5204,65 @@ func (g *GRPCServer) GetHeadlessAuthentication(ctx context.Context, req *proto.G
 	return authReq, trace.Wrap(err)
 }
 
+// WatchPendingHeadlessAuthentications watches the backend for pending headless authentication requests for the user.
+func (g *GRPCServer) WatchPendingHeadlessAuthentications(_ *emptypb.Empty, stream proto.AuthService_WatchPendingHeadlessAuthenticationsServer) error {
+	auth, err := g.authenticate(stream.Context())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	filter := types.HeadlessAuthenticationFilter{
+		Username: auth.context.User.GetName(),
+		State:    types.HeadlessAuthenticationState_HEADLESS_AUTHENTICATION_STATE_PENDING,
+	}
+
+	watcher, err := auth.NewWatcher(stream.Context(), types.Watch{
+		Name: auth.User.GetName(),
+		Kinds: []types.WatchKind{{
+			Kind:   types.KindHeadlessAuthentication,
+			Filter: filter.IntoMap(),
+		}},
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer watcher.Close()
+
+	// The headless login process may be waiting for the user to create a stub to authorize the insert.
+	// Create a stub and re-create it each time it expires.
+	if err := auth.CreateHeadlessAuthenticationStub(stream.Context()); err != nil {
+		return trace.Wrap(err)
+	}
+
+	ticker := time.NewTicker(defaults.CallbackTimeout)
+	defer ticker.Stop()
+
+	for {
```

Comment:
This feels like way too much business logic to be in `grpcserver` IMO.

Reply:
Which part? Most of this is copied over from `WatcherEvents`, which parses the watch type, creates a watcher, and streams it to through the grpc stream. The only additional parts are:
 - Preparing a `HeadlessAuthenticationFilter` using the username from the user's cert
 - Maintaining a headless stub for the lifetime of the watcher / request
   - I'm not sure what better place there is to do this, as this is logic is specifically for this request. This is similarly handled by `GetHeadlessAuthentication` in `grpcserver`.

Reply:
Is 07aa2ced8ab0c97e0ee6678910e8bc16351f6017 any better IYO?